File: test/notifications/notification_identity_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:talky_flutter/core/services/notifications/notification_identity.dart';

void main() {
  test('notificationIdForConversation is stable', () {
    expect(
      NotificationIdentity.notificationIdForConversation(42),
      NotificationIdentity.notificationIdForConversation(42),
    );
    expect(
      NotificationIdentity.notificationIdForConversation(1),
      isNot(NotificationIdentity.notificationIdForConversation(2)),
    );
    expect(
      NotificationIdentity.notificationIdForConversation(42) & 0x80000000,
      0,
    );
  });

  test('tagForConversation format', () {
    expect(NotificationIdentity.tagForConversation(99), 'conv_99');
  });

  // Mêmes vecteurs que NotificationIdentityTest (Kotlin) : si les deux hashs
  // divergent, l'annulation croisée natif ↔ Flutter (tag + id) casse en
  // silence — la notification postée par l'un devient inatteignable par
  // l'autre.
  test('vecteurs partagés avec le natif Android', () {
    expect(NotificationIdentity.notificationIdForConversation(1), 873244444);
    expect(NotificationIdentity.notificationIdForConversation(2), 923577301);
    expect(NotificationIdentity.notificationIdForConversation(42), 132351363);
    expect(NotificationIdentity.notificationIdForConversation(99), 352233207);
    expect(
      NotificationIdentity.notificationIdForConversation(123456),
      429242026,
    );
  });
}

File: android/app/src/test/kotlin/com/example/talky_flutter/NotificationIdentityTest.kt
package com.example.talky_flutter

import org.junit.Assert.assertEquals
import org.junit.Test

/**
 * Verrouille le hash FNV-1a 31 bits contre son pendant Dart
 * (`NotificationIdentity.notificationIdForConversation`,
 * `test/notifications/notification_identity_test.dart` porte les MÊMES
 * vecteurs). Si les deux divergent, l'annulation croisée des notifications
 * (tag + id) entre le natif et Flutter casse en silence.
 */
class NotificationIdentityTest {

    @Test
    fun `vecteurs partages avec Dart`() {
        assertEquals(873_244_444, MessageNotificationHelper.notificationIdForConversation(1))
        assertEquals(923_577_301, MessageNotificationHelper.notificationIdForConversation(2))
        assertEquals(132_351_363, MessageNotificationHelper.notificationIdForConversation(42))
        assertEquals(352_233_207, MessageNotificationHelper.notificationIdForConversation(99))
        assertEquals(429_242_026, MessageNotificationHelper.notificationIdForConversation(123456))
    }

    @Test
    fun `id invalide = 0`() {
        assertEquals(0, MessageNotificationHelper.notificationIdForConversation(0))
        assertEquals(0, MessageNotificationHelper.notificationIdForConversation(-5))
    }

    @Test
    fun `toujours positif sur 31 bits`() {
        for (id in intArrayOf(1, 7, 1000, Int.MAX_VALUE)) {
            val hash = MessageNotificationHelper.notificationIdForConversation(id)
            assertEquals(0, hash and Int.MIN_VALUE)
        }
    }
}
